apps/music/js/metadata/album_art_cache.js

```diff
@@ -38,17 +96,22 @@ var AlbumArtCache = (function() {
 
     // See if we've already made a URL for this album. If so, return it.
     var cacheKey = makeCacheKey(fileinfo);
-    if (cacheKey && cacheKey in L1Cache) {
-      return Promise.resolve(L1Cache[cacheKey]);
+    if (cacheKey && cacheKey in L1Cache.thumbnail) {
+      return Promise.resolve(L1Cache.thumbnail[cacheKey]);
     }
 
     // Otherwise, see if we've saved a blob in asyncStorage. If not, create a
     // thumbnail blob and store it in the cache. Finally, create a URL for the
     // blob and return it in the Promise.
+    //
+    // XXX: There's a (minor) race condition here. If two requests for the same
+    // album art get here, the slower one will overwrite the faster one's cached
+    // blob URL, meaning that not all instances of a piece of album art will
+    // have the same URL. This only really matters to tests, though.
```

Comment:
As soon as you notice that the L1 cache is empty you could store a promise in it. Then when you've got the thumbnail swap the promise out for the URL. And resolve the promise to the thumbnail as well.

This seems like it would add a lot of complexity, though.  The consequences for the race condition are minor so this seems acceptable to me as it is.  Unless there is some common case (displaying the tracks on an album) where you make many requests for the same thumbnail at the same time. Then you might want to think about it.

Actually, another easier thing you could do here is modify makeURL so that it checks again that the cache is still empty. If, by the time that makeURL is called there is already a url stored, then it could discard the new blob and return the cached URL.

That would solve it, wouldn't it? You don't have any cases where you purposely want to overwrite the cache, do you?

I think this kind of double-checking is actually a thing in concurrent programming and there might even be a computer-sciencey word for it.